### DISCUSSION TITLE
Best way to incrementally move to typescript

### DISCUSSION THREAD

**Discussion by schlaegerz** [meteor/meteor#12521] [Feature Requests]
I have been using Meteor for a while in a fairly big project and I'm looking to move it over to typescript.

My project is fairly big and I've been using flow-type https://github.com/facebook/flow but typescript has some additional benefits that would be nice to utilize. 

I am trying to move this stuff over one at a time just because all at once would be too much.

I have tried a few of the different packages and can't quite figure out how to make them work with only compiling a few files

if I try the normal typescript package it doesn't read the tsconfig so I can't make my dynamic paths work which is a must for my repo

adornis:typescript seemed to be the closest to me, but I could not get babel-module-resolver to work, and it had some other issues as it has been deprecated for a few years

refapp:meteor-typescript has the problem that is trying to run the ts compiler on all of my files which complains because of all my flow annotations and its generating the files in the source folder so its really cluttering my git with generated files.

It would be great if there was an easy way to have just a few files be typescript so that we can slowly convert the whole project. If there is a solid way to do it could some one point me to some resources?

**Comment by hexsprite** [comment]
I used to use a mix of JS+Flow on a project but since switched to a mix of JS+TS using the built-in Meteor Typescript compiler. No issues incrementally converting. 

Here's my tsconfig.json if that is useful to you. Not sure if it's all best practices but it works for me :)

```
{
  "compilerOptions": {
    /* Basic Options */
    "target": "es2018",
    "module": "esNext",
    "lib": ["esnext", "dom"],
    "allowJs": true,
    "allowSyntheticDefaultImports": true,
    "checkJs": false,
    "jsx": "preserve",
    "noEmit": true,
    "experimentalDecorators": true,
    "emitDecoratorMetadata": true,
    "skipLibCheck": true,
    /* Strict Type-Checking Options */
    "strict": true,
    "noImplicitAny": true,
    "strictNullChecks": true,

    /* Additional Checks */
    "noUnusedLocals": true,
    "noUnusedParameters": true,
    "noImplicitReturns": false,
    "noFallthroughCasesInSwitch": false,
    "isolatedModules": true,
    /* Module Resolution Options */
    "baseUrl": ".",
    "paths": {
      /* Support absolute /imports/* with a leading '/' */
      "/imports/*": ["imports/*"],
      "/server/*": ["server/*"],
      "/client/*": ["client/*"],
      "/lib/*": ["lib/*"],
      "meteor/*": [
        "node_modules/@types/meteor/*"
        // ".meteor/local/types/packages.d.ts"
      ],
      "cordova/cordova-plugin-purchases": [
        ".meteor/local/cordova-build/node_modules/cordova-plugin-purchases/www/plugin.d.ts"
      ]
      // "*": ["*", ".meteor/local/cordova-build/node_modules/*"]
    },
    "moduleResolution": "node",
    "resolveJsonModule": true,
    "types": ["node", "mocha", "jest"],
    "preserveSymlinks": true,
    "typeRoots": ["./typings"]
  },
  "exclude": [
    ".build",
    ".meteor/local/.build*",
    ".meteor/local/.build*/**",
    "devops",
    "packages",
    "private"
  ]
}
```

**Comment by schlaegerz** [comment]
Hmm thanks for the reply. I must be missing some other step or not understanding something.

If I just have the meteor typescript package and change one file to .ts 

I get a file not found error in my browser, and if it try to change my main module 
```

  "meteor": {
    "mainModule": {
      "client": "client/lib/init.ts",
```
I get 
error: Could not find mainModule for 'web.browser' architecture: client/lib/init.ts

This is my versions
```
accounts-base@2.2.6
accounts-facebook@1.3.3
accounts-oauth@1.4.2
accounts-password@2.3.3
aldeed:collection2@3.5.0
allow-deny@1.1.1
autoupdate@1.8.0
babel-compiler@7.10.2
babel-runtime@1.5.1
base64@1.0.12
binary-heap@1.0.11
blaze-tools@1.1.3
boilerplate-generator@1.7.1
caching-compiler@1.2.2
caching-html-compiler@1.2.1
callback-hook@1.5.0
cfs:graphicsmagick@0.0.18
check@1.3.2
cultofcoders:redis-oplog@2.1.1
ddp@1.4.1
ddp-client@2.6.1
ddp-common@1.4.0
ddp-rate-limiter@1.1.1
ddp-server@2.6.0
diff-sequence@1.1.2
disable-oplog@1.0.7
dynamic-import@0.7.2
ecmascript@0.16.5
ecmascript-runtime@0.8.0
ecmascript-runtime-client@0.12.1
ecmascript-runtime-server@0.11.0
ejson@1.1.3
email@2.2.3
es5-shim@4.8.0
facebook-oauth@1.11.2
fetch@0.1.3
fortawesome:fontawesome@4.7.0
fourseven:scss@4.15.0
gadicohen:robots-txt@0.0.10
gadicohen:sitemaps@0.0.26
geojson-utils@1.0.11
hot-code-push@1.0.4
hot-module-replacement@0.5.2
html-tools@1.1.3
htmljs@1.1.1
http@2.0.0
id-map@1.1.1
inter-process-messaging@0.1.1
launch-screen@1.3.0
lmieulet:meteor-coverage@4.1.0
localstorage@1.2.0
logging@1.3.1
meteor@1.11.0
meteor-base@1.5.1
meteortesting:browser-tests@1.5.1
meteortesting:mocha@2.0.4
meteortesting:mocha-core@8.1.2
minifier-css@1.6.2
minifier-js@2.7.5
minimongo@1.9.1
mobile-experience@1.1.0
mobile-status-bar@1.1.0
modern-browsers@0.1.9
modules@0.19.0
modules-runtime@0.13.1
modules-runtime-hot@0.14.1
mongo@1.16.4
mongo-decimal@0.1.3
mongo-dev-server@1.1.0
mongo-id@1.0.8
npm-mongo@4.12.1
oauth@2.1.3
oauth2@1.3.2
ordered-dict@1.1.0
ostrio:cookies@2.7.2
ostrio:files@2.3.2
percolate:migrations@1.1.0
promise@0.12.2
raix:eventemitter@1.0.0
random@1.2.1
rate-limit@1.0.9
react-fast-refresh@0.2.5
react-meteor-data@2.6.1
reactive-var@1.0.12
reload@1.3.1
retry@1.1.0
routepolicy@1.1.1
seba:minifiers-autoprefixer@2.0.1
service-configuration@1.3.1
sha@1.0.9
socket-stream-client@0.5.0
spacebars-compiler@1.3.1
standard-minifier-js@2.8.1
static-html@1.3.2
templating-tools@1.2.2
tmeasday:acceptance-test-driver@0.1.0
tmeasday:check-npm-versions@1.0.2
tracker@1.3.0
typescript@4.7.4
underscore@1.0.11
url@1.3.2
webapp@1.13.3
webapp-hashing@1.1.1
zeroasterisk:throttle@0.3.5
```

and my .babelrc

```
{
  "assumptions":{
     "superIsCallableConstructor":false
  },
  "presets":[
     "@babel/preset-flow",
     [
        "@babel/preset-env",
        {
           "targets":">1%"
        }
     ],
     "@babel/react",
     ["@babel/preset-typescript", {"allowNamespaces": true}]
  ],
  "plugins":[
     "@babel/transform-runtime",
     [
        "module-resolver",
        {
           "root":[
              "./"
           ],
           "extensions":[ ".js", ".jsx"],
           "alias":{
              "^react-native$":"./node_modules/react-native-web",
              "@shared":"./common"
           }
        }
     ],
     "@babel/plugin-syntax-flow",
     "@babel/plugin-transform-flow-strip-types"
  ]
}
```

And my understanding is the base package doesn't actually parse the tsconfig? But maybe I misunderstand it

**Comment by hexsprite** [comment]
Yeah, you're right, I only needed the tsconfig for VSCode usage.

As for your compilation problem, have you creating a minimal typescript starter project based on the builtin templates ? And then seeing if you can rename a js to a ts?

**Comment by chneau** [comment]
If that's of any help, I've created a template for meteor projects here https://github.com/chneau/meteor-template
I've created a class to handle strongly typed meteor methods call and meteor subscribes calls you can find here https://github.com/chneau/meteor-template/tree/master/imports 

This allows declaration of methods this way 

```ts
export const mutateLinkUpdatedAt = new TypedMethod({
  name: "links.mutateLinkUpdatedAt",
  async run(linkId: string) {
    return await LinksCollection.updateAsync(linkId, { $set: { updatedAt: new Date() } });
  },
});
```

And it's making life easier when importing this method from another file, calling it this way: `mutateLinkUpdatedAt.call("linkId");`  whith typescript whining if the parameters are not properly set.

**Comment by schlaegerz** [comment]
Ok this was problem with my local setup. I thought it was more general problem than what it ended up being
